### PR TITLE
Extend aggregate_ping renderer

### PR DESCRIPTION
**Pull request by ioc32**
It still needs more love (better formatting, testing...), but at least it works with msm#1001:
```
$ ripe-atlas report --renderer aggregate_ping 1001

RIPE Atlas Report for Measurement #1001
===================================================
Collecting results...
-- 193.0.14.129 ping statistics ---
26711 packets transmitted, 26350 received, 1.35150312605% loss
rtt min/med/avg/max = 0.209/33.274/60.5296812673/14027.879 ms
```
We'll rebase all the commits in this PR before merging.

Also, how about adding this to the standard ping renderer?

### CODE REVIEW

**Comment by astrikos** [comment]
I vote for adding it to the standard one, it's nice to see it by default and not having to specify a diff aggregator.

**Comment by pierky** [comment]
Same for me, it would be nice to have in the std ping. I also suggest to add a counter of how many probes received and didn't receive a response.

**Comment by emileaben** [comment]
I don't understand what the output is.
aggregate results for *all probes* for some time period?

If it's that, I'm not sure how the numbers coming out of this help me?

What it says is that somewhere there is a probe with really low RTT, there
also is a probe with really high RTT, and out of this there is packet loss
somewhere.
Because RIPE atlas is large and diverse, that's going to be the output
always if you have a significant number of probes.

**Comment by ioc32** [comment]
For some reason the reply I left over email has not made it to the public discussion.
Copy pasting it here:

> An Internet Data Junkie does ;-) but one doesn't always necessarily need to use a large number of probes. I'm thinking of small to medium LIRs and how could they use this tool, or how this tool could be useful to them. 

>Admittedly the usefulness of this summary stats rapidly drops as the number of probes involved in the measurements increases.

>For those, I think we'll need to leverage the aggregators that are being worked on atm. 
 
>All in all, I believe mimicking standard ping's output is worthwhile even if it has some caveats that should be taken into account when interpreting the results.

**Comment by ioc32** [comment]
I've also tested it with other measurements (including a really old one from 2011).
Other than ripe.atlas.sagan complaining about Unknown Firmwares (version=0) on stderr, the report looks good.

@emileaben we can merge this in aggregate_ping as an example and decide if we want to add it to the standard ping later

@danielquinn if you're curious about that old msm, it's 1000047

**Comment by ioc32** [comment]
@pierky regarding the addition of counters, that's a good suggestion!

I guess we can track lost packets per probe and print the number of probes with a packet loss above a given threshold?

**Comment by pierky** [comment]
> I guess we can track lost packets per probe and print the number of probes with a packet loss above a given threshold?

:+1:

**Comment by ioc32** [comment]
@pierky meh, I gave it a try [1] but the results are not all that useful since the reports are by default generated with the measurement-latest API, so you only get, say, 3 packets per result.

Thresholds do not quite match the possible packet loss rates per probe in this case (0/3, 1/3, 2/3, 3/3). We'd need to build reports on results from a larger interval for these thresholds to make sense.

[1] https://github.com/ioc32/ripe-atlas-tools/tree/agg-stats

File: ripe/atlas/tools/renderers/aggregate_ping.py [hunk 1]
```diff
@@ -11,22 +11,58 @@ class Renderer(BaseRenderer):
...
-            **self.totals
+        self.packet_loss = (1 - float(self.totals["received"]) / self.totals["sent"]) * 100
+        return self.render(
+            "reports/aggregate_ping.txt",
+            target=self.target,
+            sent=self.totals["sent"],
+            received=self.totals["received"],
+            packet_loss=self.packet_loss,
+            min=min(min for min in self.totals["min"] if min is not None),
+            med=self.median(self.totals["rtts"]),
```
Comment (danielquinn) [inline ripe/atlas/tools/renderers/aggregate_ping.py]: This all looks fine by me, but could you please use `median` and `mean` instead of `med` and `avg`?  It's obvious what's going on there, but in the template where these values are used, 3 letters isn't enough.

Comment (ioc32) [inline ripe/atlas/tools/renderers/aggregate_ping.py]: Sure, np.

On Wed, Sep 30, 2015 at 12:15 PM, Daniel Quinn <notifications@github.com>
wrote:

> In ripe/atlas/tools/renderers/aggregate_ping.py
> <https://github.com/RIPE-NCC/ripe-atlas-tools/pull/22#discussion_r40778365>
> :
>
> >
> >          return ""
> >
> >      def on_finish(self):
> > -        return "\nTotals:\n  Sent: {sent}\n  Received: {received}\n".format(
> > -            **self.totals
> > +        self.packet_loss = (1 - float(self.totals["received"]) / self.totals["sent"]) * 100
> > +        return self.render(
> > +            "reports/aggregate_ping.txt",
> > +            target=self.target,
> > +            sent=self.totals["sent"],
> > +            received=self.totals["received"],
> > +            packet_loss=self.packet_loss,
> > +            min=min(min for min in self.totals["min"] if min is not None),
> > +            med=self.median(self.totals["rtts"]),
>
> This all looks fine by me, but could you please use median and mean
> instead of med and avg? It's obvious what's going on there, but in the
> template where these values are used, 3 letters isn't enough.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/RIPE-NCC/ripe-atlas-tools/pull/22/files#r40778365>.
>



-- 
"If you want to go fast, go alone. If you want to go far, go together."

Comment (astrikos) [inline ripe/atlas/tools/renderers/aggregate_ping.py]: @ioc32 could you make the change so we can merge and close? the more we leave it open the more risky is to have someone else write on the same file and create conflict :)

**Comment by ioc32** [comment]
I renamed the variables as per Dan's request. Since this aggregate_ping renderer is likely meant to be more of a "code reference" rather than the renderer operators are going to use, I am happy to merge and close.

@emileaben has very reasonable concerns about the PR, so we should go through those before changing the base ping renderer, IMHO.
